### PR TITLE
Test OTX segmentation models

### PR DESCRIPTION
**Pull request by Wovchena**
I have a feeling that comparing just unique values for segmentation is a weak test. I will update it later to compare histograms

### CODE REVIEW

File: docs/model-configuration.md [hunk 1]
```diff
@@ -57,6 +57,9 @@ The list features only model wrappers which intoduce new configuration values in
 #### `SegmentationModel` and its subclasses
 1. `labels`: List - list of class labels
 1. `path_to_labels`: str - path to file with labels. Overrides the labels, if they sets via 'labels' parameter
+1. `blur_strength`: int - blurring kernel size. -1 signals not to apply bluring and soft_threshold
```
Comment (AlexKoff88) [inline docs/model-configuration.md:60]: `bluring` -> `blurring` everywhere.
`retern` -> `return`
`signals not to apply` sounds not well. I would rephrase to, e.g. `-1 value means no blurring and ...`

File: model_api/cpp/models/src/segmentation_model.cpp [hunk 2]
```diff
@@ -32,8 +32,80 @@
 #include "models/input_data.h"
 #include "models/results.h"
 
+namespace {
+cv::Mat create_hard_prediction_from_soft_prediction(const cv::Mat& soft_prediction, float soft_threshold, int blur_strength) {
+    cv::Mat soft_prediction_blurred;
```
Comment (AlexKoff88) [inline model_api/cpp/models/src/segmentation_model.cpp:37]: @sovrasov, please take a look as well.

File: model_api/cpp/models/src/segmentation_model.cpp [hunk 3]
```diff
@@ -32,8 +32,80 @@
...
+        if (model->has_rt_info("model_info", "soft_threshold")) {
+            soft_threshold = stof(model->get_rt_info<std::string>("model_info", "soft_threshold"));
+        }
+    } else {
+        soft_threshold = soft_threshold_iter->second.as<float>();
+    }
+    auto return_soft_prediction_iter = configuration.find("return_soft_prediction");
+    if (return_soft_prediction_iter == configuration.end()) {
+        if (model->has_rt_info("model_info", "return_soft_prediction")) {
+            return_soft_prediction = model->get_rt_info<std::string>("model_info", "return_soft_prediction") == "True";
```
Comment (AlexKoff88) [inline model_api/cpp/models/src/segmentation_model.cpp:85]: There is a conversion to bool tape as well as far as I remember. So, no need to compare strings.

File: tests/python/accuracy/public_scope.json [hunk 4]
```diff
@@ -6,7 +6,55 @@
...
+        ]
+    },
+    {
+        "name": "otx_models/Lite-hrnet-18.xml",
+        "type": "SegmentationModel",
+        "test_data": [
+            {
+                "image": "coco128/images/train2017/000000000074.jpg",
+                "reference": [
+                    "0 1 "
```
Comment (AlexKoff88) [inline tests/python/accuracy/public_scope.json:21]: I agree that this is not enough and should be enhanced. Normally, we should take references from OTX, and save them here. Maybe @sovrasov can help us to get such references. It should not take a lot of time. It's just a processed output 
 for one sample.

File: docs/model-configuration.md (hunk 1 above)
Comment (Wovchena) [reply docs/model-configuration.md]: Applied

File: model_api/cpp/models/src/segmentation_model.cpp (hunk 3 above)
Comment (AlexKoff88) [reply model_api/cpp/models/src/segmentation_model.cpp:85]: Ok, got it. OpenVINO is very separate as usual here in terms of representing the things )

File: model_api/cpp/models/src/segmentation_model.cpp [hunk 5]
```diff
@@ -32,8 +32,80 @@
 #include "models/input_data.h"
 #include "models/results.h"
 
+namespace {
+cv::Mat create_hard_prediction_from_soft_prediction(const cv::Mat& soft_prediction, float soft_threshold, int blur_strength) {
```
Comment (sovrasov) [inline model_api/cpp/models/src/segmentation_model.cpp:36]: Why soft_threshold argument is not used here? Perhaps, thresholding makes no sense in most cases, since we have argmax, but in python you keep it.

File: model_api/cpp/models/src/segmentation_model.cpp [hunk 6]
```diff
@@ -32,8 +32,80 @@
...
+    cv::Mat soft_prediction_blurred;
+    cv::blur(soft_prediction, soft_prediction_blurred, cv::Size{blur_strength, blur_strength});
+    assert(soft_prediction_blurred.channels() > 1);
+    cv::Mat hard_prediction{cv::Size{soft_prediction_blurred.rows, soft_prediction_blurred.cols}, CV_8UC1};
+    for (int i = 0; i < soft_prediction_blurred.rows; ++i) {
+        for (int j = 0; j < soft_prediction_blurred.cols; ++j) {
+            float max_prob = -std::numeric_limits<float>::infinity();
+            uint8_t max_id = 0;
+            for (int c = 0; c < soft_prediction_blurred.channels(); ++c) {
+                float prob = ((float*)soft_prediction_blurred.data)[
```
Comment (sovrasov) [inline model_api/cpp/models/src/segmentation_model.cpp:46]: Here Mat.ptr(i) could be used to reduce the indexing expression

File: tests/python/accuracy/public_scope.json (hunk 4 above)
Comment (sovrasov) [reply tests/python/accuracy/public_scope.json:21]: As soon as MAPI is a standalone tool, I'd propose not to tie to references from OTX. Instead, we may check C++ vs python output. Inconsistencies with OTX could be fixed once during the integration and later validation pytorch task vs openvino task will keep the OV wrappers in shape.

File: model_api/cpp/models/src/segmentation_model.cpp (hunk 5 above)
Comment (Wovchena) [reply model_api/cpp/models/src/segmentation_model.cpp:36]: Yes, that's strange. It makes sense to return the most probable label, not the one with the smallest index if every labels has low probability for a given pixel. Can I completely drop soft_prediction? But this will introduce difference between OTX and ModelAPI

File: model_api/cpp/models/src/segmentation_model.cpp (hunk 6 above)
Comment (Wovchena) [reply model_api/cpp/models/src/segmentation_model.cpp]: Updated

File: model_api/cpp/models/src/segmentation_model.cpp [hunk 7]
```diff
@@ -32,8 +32,80 @@
...
+            float max_prob = -std::numeric_limits<float>::infinity();
+            uint8_t max_id = 0;
+            for (int c = 0; c < soft_prediction_blurred.channels(); ++c) {
+                float prob = ((float*)soft_prediction_blurred.ptr(i, j))[c];
+                if (prob > max_prob) {
+                    max_prob = prob;
+                    max_id = c;
+                }
+            }
+            hard_prediction.at<uint8_t>(i, j) = max_id;
```
Comment (sovrasov) [inline model_api/cpp/models/src/segmentation_model.cpp:52]: The result of discussion: set hard prediction to 0 if max_prob < soft_threshold

Comment (Wovchena) [reply model_api/cpp/models/src/segmentation_model.cpp:52]: Updated

**Review by AlexKoff88 — APPROVED** [review]

**Comment by AlexKoff88** [comment]
Let's merge.
